src/djinterop/enginelibrary/el_track_impl.hpp

```diff
@@ -135,6 +135,17 @@ class el_track_impl : public djinterop::track_impl
     template <typename T>
     void set_perfdata(const char* column_name, const T& content)
     {
+        auto encoded_content = content.encode();
+        // Check that subsequent reads can correctly decode what we are about to
+        // write.
+        if (!(T::decode(encoded_content) == content))
+        {
+            throw std::logic_error{
+                "Data supplied for column " + std::string(column_name) +
+                " is not invariant under encoding and subsequent decoding. "
```

Comment:
This PR allows to read beat data if the beatgrid is malformed. If the beatgrid is malformed, then it is just discarded. When doing such a step, it is important that we don't accidentally write a malformed beatgrid because then it would be silently discarded in subsequent reads. Therefore I implemented this check. The error is only triggered when we have a programming error (i.e. a bug in libdjinterop) which I also stated in the error mesage.

Reply:
Ok, but have you encountered any real-world situations with malformed beatgrid data?

If this is a common occurrence, perhaps we should write a unit test for it (add a `TODO` if you like).

Reply:
I was wrong about that it is necessarily a bug in libdjinterop when the error is triggered. It might also be triggered if the library user attempts to write a malformed beat grid. Previously in that case the beat grid would just be written to the database and afterwards sampling and beat grids could neither be written nor read using libdjinterop, because we have checks in `beat_data::decode()`.

With just the change in `src/djinterop/enginelibrary/performance_data_format.cpp` the error is removed (and there's a TODO that it's a warning instead), so you can read old libraries that have already been corrupted by the above mentioned error.

However, this usage error should not just trigger a warning when reading, but also an error when writing. The roundtrip check achieves just that.

Alternatively we could duplicate all of the checks we do when decoding and also perform them when encoding, but this would be code duplication with massive potential for bugs where an edge case still slips through the encoding checks.

You might think that this roundtrip check does not give good error messages because it's just a default message. But currently most of the potential errors are thrown inside `T::decode(encoded_content)` and have their own error message. As we become less and less pedantic when decoding (so that we can read from more kinds of corrupted databases), those errors go away and there's just the default error message “Data [...] is not invariant under encoding and subsequent decoding”. But then there should be a warning inside the `decode()` function, so whenever you get the cited default error message, you can just look which warning(s) you got beforehand.

The warning in `beat_data::decode()` is not implemented, yet, because we didn't agree on how we handle warnings, yet

Reply:
I think the overhead of this check is small, because in `!(T::decode(encoded_content) == content)` there's not a single database access.